examples/ecs/archetype_invariants.rs

```diff
@@ -0,0 +1,84 @@
+//! Archetype invariants are rules about which combinations of components can coexist.
+//!
+//! An archetype (in the sense that Bevy uses it) is the "unique set of components" that belong to an entity.
+//! These are useful to codify your assumptions about the composition of your entities.
+//! For example, an entity can never have a `Player` component with a `Camera` together,
+//! or a `GlobalTransform` may only be valid in association with a `Transform`.
+//! By constructing `ArchetypeInvariant`s out of `ArchetypeStatement`s,
+//! we can encode this logic into our app.
+//!
+//! Archetype invariants are guaranteed to hold at *all* points during the app's lifecycle;
+//! this is automtically checked on component insertion and removal, including when entities are spawned.
+//! Make sure to test thoroughly when using archetype invariants in production though;
+//! any violations will result in a panic!
+//!
+//! There are many helper methods provided on `ArchetypeInvariant` to help easily construct common invariant patterns,
+//! but we will only be showcasing some of them here.
+//! For a full list, see the docs for `ArchetypeInvariant`.
+use bevy::prelude::*;
+
+fn main() {
+    App::new()
+        // Archetype invariants are constructed in terms of bundles;
+        // use (MyComponent, ) to construct a bundle with a single item.
+        // This invariant ensures that Player and Camera can never be found together.
+        .add_archetype_invariant(ArchetypeInvariant::<(Player,), (Camera,)>::forbids())
```

Comment:
Yeah, I'm not sure if the builder pattern would be better here.

For your second question, yes, we're blocked on #2975 🥺

Reply:
Not really builder pattern- just some other type that takes one type parameter and all its methods take the other. 
Or even spookier, implement this methods as traits on bundles themselves.
`(Player,)::forbids::<(Camera,)>()`
Assuming we get #2975 in, that's `Player::forbids::<Camera>()`. Much more readable imo. And I suppose it doesn't really "pollute" the bundles' autocomplete space since components and bundles don't tend to have methods?
That one I can see being even more controversial than a helper type though.

Reply:
Definitely agreed that the readability could be better here. I think that #2975 is probably the highest-leverage piece of work to improve this. For me, just going from `ArchetypeInvariant::<(Player,), (Camera,)>::forbids()` to `ArchetypeInvariant::<Player, Camera>::forbids()` is a serious improvement. 

As for the infix stuff, I personally think it's less readable than the current API, but I think this is mostly a matter of personal preference. I'd like to get others' thoughts on this before we make a decision one way or the other.

Reply:
After [some discussion on discord](https://discord.com/channels/691052431525675048/996129452734029824), it looks like something like `Player::forbids::<Camera>()` is the winning syntax, though @cart prefers a trait-based approach, where the helper functions become helper types implementing an invariant trait. I'm going to play around with the trait based approach, but if it ends up too messy, then I'll go with the `Player::forbids::<Camera>()` approach.

Reply:
Implemented as of 21a1a2350d08982d5e86293f4d43db02ed8146a2